tinycrypt/ecc.c

```diff
@@ -266,15 +272,76 @@ static void muladd(uECC_word_t a, uECC_word_t b, uECC_word_t *r0,
 
 }
 
-/* Computes result = left * right. Result must be 2 * num_words long. */
-static void uECC_vli_mult(uECC_word_t *result, const uECC_word_t *left,
-			  const uECC_word_t *right, wordcount_t num_words)
+/* State for implementing random delays in uECC_vli_mult_rnd().
+ *
+ * The state is initialised by randomizing delays and setting i = 0.
+ * Each call to uECC_vli_mult_rnd() uses one byte of delays and increments i.
+ *
+ * A scalar muliplication uses 14 field multiplications per bit of exponent.
+ */
+typedef struct {
+	uint8_t delays[14 * 256];
+	uint16_t i;
+} wait_state_t;
+
+#if 0
+static void wait_init(wait_state_t *s)
+{
+	g_rng_function(s->delays, sizeof(s->delays));
+	s->i = 0;
+}
+#endif
+
+/* Computes result = left * right. Result must be 2 * num_words long.
+ *
+ * As a counter-measure against horizontal attacks, add noise by performing
+ * a random number of extra computations performing random additional accesses
+ * to limbs of the input.
+ *
+ * Each of the two actual computation loops is surrounded by two
+ * similar-looking waiting loops, to make the beginning and end of the actual
+ * computation harder to spot.
+ *
+ * We add 4 waiting loops of between 0 and 3 calls to muladd() each. That
+ * makes an average of 6 extra calls. Compared to the main computation which
+ * makes 64 such calls, this represents an average performance degradation of
+ * less than 10%.
+ *
+ * Compared to the original uECC_vli_mult(), loose the num_words argument as we
+ * know it's always 8. This saves a bit of code size and execution speed.
+ */
+static void uECC_vli_mult_rnd(uECC_word_t *result, const uECC_word_t *left,
+			      const uECC_word_t *right, wait_state_t *s)
 {
 
 	uECC_word_t r0 = 0;
 	uECC_word_t r1 = 0;
 	uECC_word_t r2 = 0;
 	wordcount_t i, k;
+	uint8_t delays;
+	uECC_word_t rr0 = 0, rr1 = 0;
+	volatile uECC_word_t r;
+	const uint8_t num_words = 8;
+
+	if (s) {
+		/* Fetch 8 bit worth of random delays from the state */
+		delays = s->delays[s->i++];
```

Comment:
Can `s->i++` be incremented over table size?

Reply:
According to the following reasoning it can't:

1. Each call to `uECC_vli_mult_rnd()` increments `i` exactly once and no other function directly increments `i`.
2. Each call to `XYcZ_add()` calls `uECC_vli_mult_rnd()` exactly 6 times (via `modMult` or `modSquare`), and each call to `XYcZ_addC()` calls it exactly 8 times.
3. So each pair of calls to `XYcZ_add()` and `XYcZ_addC()` increments `i` exactly 14 times.
4. In all calls to `EccPoint_mult()` the `num_bits` arguments is `curve->num_n_bits + 1` which is always 257. So the main `for` loop it that function execute 255 times, with the loop body incrementing `i` 14 times, and there's an extra pair of calls to `XYcZ_add()` and `XYcZ_addC()` outside the loop, and not other (indirect) calls to `uECC_vli_mult_rnd()` in that function, resulting in `i` being incremented 14 * 256 times in that function.
5. Since `i` is set to `0` at the beginning of the function, is incremented after being used, and the table has exactly 14 * 256 elements, we should be safe :)

I'll add a comment to that effect near the definition of the table size and the increment.

Reply:
Note: this was also confirmed experimentally by printing out the value of `i` right before `EccPoint_mult()` returns.

Reply:
Wow, I just realised I did something pretty stupid: I wanted to save on execution speed by only calling the RNG once, so I put all the state in one table, but that makes that table very large: 3584 bytes. Allocating such a big table is really not a good idea (especially on the stack). I'll change the code to use only 14 bytes of delays at once, and reset the table before each pair of calls to `XYcZ_add()` and `XYcZ_addC()`. As a welcome side-effect, this should make the reasoning as to why `i` doesn't overflow much easier.

Reply:
Done.